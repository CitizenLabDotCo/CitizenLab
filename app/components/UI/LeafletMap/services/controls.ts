import L from 'leaflet';

export function addLayersControl(map: L.Map) {
  return L.control.layers().addTo(map);
}

export function removeControl(
  map: L.Map,
  layerControl?: L.Control.Layers | null
) {
  if (layerControl) {
    map.removeControl(layerControl);
  }
}
