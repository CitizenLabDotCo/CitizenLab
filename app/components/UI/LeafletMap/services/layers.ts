import {
  isNilOrError,
  isFunction,
  isString,
  isOrReturnsString,
} from 'utils/helperUtils';
import L from 'leaflet';
import { isEmpty, cloneDeep, reverse } from 'lodash-es';

import { markerIcon } from './markers';

import {
  GeoJSONLayer,
  IMarkerStringOrObjectOrFunctionForLayer,
  IOverlayStringOrObjectOrFunctionForLayer,
  ITooltipStringOrObjectOrFunctionForLayer,
  IPopupStringOrObjectOrFunctionForLayer,
} from '../typings';

import { DEFAULT_MARKER_ICON } from '../config';

export function addPopupToLayer(
  layer: L.GeoJSON,
  feature: GeoJSON.Feature,
  popupStringOrFunction?: IPopupStringOrObjectOrFunctionForLayer
) {
  if (!popupStringOrFunction) {
    return;
  }

  let popup: string = '';

  if (isFunction(popupStringOrFunction)) {
    popup = popupStringOrFunction(layer, feature);
  } else if (isString(popupStringOrFunction)) {
    popup = popupStringOrFunction;
  }

  if (popup && !isEmpty(popup)) {
    layer.bindPopup(popup);
  }
}

export function addTooltipToLayer(
  layer: L.GeoJSON,
  feature: GeoJSON.Feature,
  tooltipStringOrFunction?: ITooltipStringOrObjectOrFunctionForLayer
): void {
  if (!tooltipStringOrFunction) {
    return;
  }

  let tooltip: string = '';

  if (isFunction(tooltipStringOrFunction)) {
    tooltip = tooltipStringOrFunction(layer, feature);
  } else if (isString(tooltipStringOrFunction)) {
    tooltip = tooltipStringOrFunction;
  }

  if (tooltip && !isEmpty(tooltip)) {
    layer.bindTooltip(tooltip);
  }
}

export function addLayerOverlay(
  layer: L.GeoJSON,
  layerControl: L.Control.Layers,
  geoJSONLayer: GeoJSONLayer,
  overlayStringOrOptionsOrFunction: IOverlayStringOrObjectOrFunctionForLayer
): void {
  let overlayContent;

  if (isString(overlayStringOrOptionsOrFunction)) {
    overlayContent = overlayStringOrOptionsOrFunction;
  } else if (isFunction(overlayStringOrOptionsOrFunction)) {
    overlayContent = overlayStringOrOptionsOrFunction(geoJSONLayer);
  }

  if (isString(overlayContent)) {
    layerControl.addOverlay(layer, overlayContent);
  }
}

export function layerMarker(
  geojsonLayer: GeoJSONLayer,
  latlng: L.LatLng,
  markerStringOrOptionsOrFunction: IMarkerStringOrObjectOrFunctionForLayer,
  options = {}
): L.Marker {
  let marker: L.Icon;

  if (isString(markerStringOrOptionsOrFunction)) {
    marker = markerIcon({ url: markerStringOrOptionsOrFunction });
  } else if (
    isOrReturnsString(markerStringOrOptionsOrFunction, geojsonLayer, latlng)
  ) {
    marker = markerIcon({
      url: markerStringOrOptionsOrFunction(geojsonLayer, latlng),
    });
  } else if (isFunction(markerStringOrOptionsOrFunction)) {
    marker = markerIcon(markerStringOrOptionsOrFunction(geojsonLayer, latlng));
  } else {
    marker = markerIcon(markerStringOrOptionsOrFunction);
  }

  return L.marker(latlng, { ...options, icon: marker || DEFAULT_MARKER_ICON });
}

export function addLayers(
  map: L.Map,
  layersGeoJson: GeoJSONLayer[],
  { layerControl, overlay, popup, tooltip, marker }
): L.GeoJSON[] {
  return reverse(cloneDeep(layersGeoJson))
    ?.filter((layerObject) => !isEmpty(layerObject.geojson))
    .map((layerObject) => {
      const options = {
        useSimpleStyle: true,
        useMakiMarkers: true,
        pointToLayer: (_feature, latlng) => {
          layerMarker(layerObject, latlng, marker);
        },
        onEachFeature: (feature, layer) => {
          addTooltipToLayer(layer, feature, tooltip);
          addPopupToLayer(layer, feature, popup);
        },
      };

      const layer = L.geoJSON(layerObject.geojson, options as any).addTo(map);

      if (!isNilOrError(layerControl) && !isNilOrError(overlay)) {
        addLayerOverlay(layer, layerControl, layerObject, overlay);
      }

      return layer as L.GeoJSON;
    });
}

export function removeLayers(map: L.Map, leafletLayers: L.Layer[]): void {
  leafletLayers.forEach((layer) => {
    removeLayer(map, layer);
  });
}

export function removeLayer(map: L.Map, leafletLayer: L.Layer): void {
  if (leafletLayer) {
    map.removeLayer(leafletLayer);
  }
}
