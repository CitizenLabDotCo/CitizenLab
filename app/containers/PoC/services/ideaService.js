import { API_PATH } from 'containers/App/constants';
import Streams from '../utils/streams';
import Rx from 'rxjs/Rx';
import _ from 'lodash';

export function observeSingleIdea(id, localProperties = false) {
  return Streams.create(`${API_PATH}/ideas/${id}`, null, localProperties);
}

export function observeMultipleIdeas(queryParameters = null, localProperties = false) {
  return Streams.create(`${API_PATH}/ideas`, queryParameters, localProperties);
}

export function observeIdeaRelationShips(idea) {
  const { relationships } = idea;

  const topicIds = _(relationships.topics.data).map((topic) => topic.id).value();
  const areaIds = _(relationships.areas.data).map((area) => area.id).value();
  const imageIds = _(relationships.idea_images.data).map((ideaImage) => ideaImage.id).value();
  const authorId = (relationships.author.data ? relationships.author.data.id : null);
  const projectId = (relationships.author.data ? relationships.project.data.id : null);

  const topicsStreams = (topicIds.length > 0 ? Rx.Observable.combineLatest(
    topicIds.map((topicId) => Streams.create(`${API_PATH}/topics/${topicId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const areasStreams = (areaIds.length > 0 ? Rx.Observable.combineLatest(
    areaIds.map((areaId) => Streams.create(`${API_PATH}/areas/${areaId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const imagesStreams = (imageIds.length > 0 ? Rx.Observable.combineLatest(
    imageIds.map((imageId) => Streams.create(`${API_PATH}/ideas/${idea.id}/images/${imageId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const authorStream = (authorId ? Streams.create(`${API_PATH}/users/${authorId}`).observable : Rx.Observable.of(null));

  const projectStream = (projectId ? Streams.create(`${API_PATH}/projects/${projectId}`).observable : Rx.Observable.of(null));

  return Rx.Observable.combineLatest(
    topicsStreams,
    areasStreams,
    imagesStreams,
    authorStream,
    projectStream,
    (s1, s2, s3, s4, s5) => {
      return {
        topics: s1,
        areas: s2,
        images: s3,
        author: s4,
        project: s5,
      };
    }
  );
}

export function toggleIdea(observers, ideaID) {
  const action = (ideas) => ideas.map((idea) => {
    if (idea.id === ideaID) {
      return {
        ...idea,
        selected: !idea.selected,
      };
    }

    if (idea.id !== ideaID && idea.selected) {
      return {
        ...idea,
        selected: false,
      };
    }

    return idea;
  });

  observers.forEach((observer) => observer.next(action));
}
