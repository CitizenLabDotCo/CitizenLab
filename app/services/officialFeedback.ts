import { API_PATH } from 'containers/App/constants';
import streams, { IStreamParams } from 'utils/streams';
import { IRelationship, Multiloc } from 'typings';

interface IOfficialFeedbackPost {
  body_multiloc: Multiloc;
  author_multiloc: Multiloc;
  created_at: string;
  updated_at: string;
}

export interface IOfficialFeedbackData {
  id: string;
  type: 'official_feedbacks';
  attributes: IOfficialFeedbackPost;
  relationships: {
    idea: {
      data: IRelationship;
    };
    user: {
      data: IRelationship | null;
    };
  };
}

export interface IOfficialFeedback {
  data: IOfficialFeedbackData[];
}

export interface IUpdatedOfficialFeedback {
  author_id: Multiloc;
  body_multiloc: Multiloc;
}

export interface INewFeedback {
  author_multiloc: Multiloc;
  body_multiloc: Multiloc;
}

export function officialFeedbackStream(officialFeedbackId: string) {
  return streams.get<IOfficialFeedback>({ apiEndpoint: `${API_PATH}/official_feedback/${officialFeedbackId }` });
}

export function officialFeedbackForIdeaStream(ideaId: string, streamParams: IStreamParams | null = null) {
  return streams.get<IOfficialFeedback>({ apiEndpoint: `${API_PATH}/ideas/${ideaId}/official_feedback`, ...streamParams });
}

export async function addOfficialFeedbackToIdea(ideaId: string, feedBack: INewFeedback) {
  const bodyData = {
    official_feedback: feedBack
  };

  return streams.add<IOfficialFeedback>(`${API_PATH}/ideas/${ideaId}/official_feedback`, bodyData);
}

export function updateOfficialFeedback(officialFeedbackId: string, object: IUpdatedOfficialFeedback) {
  return streams.update<IOfficialFeedback>(`${API_PATH}/official_feedback/${officialFeedbackId}`, officialFeedbackId, { comment: object });
}

export function deleteOfficialfeedback(projectId: string) {
  return streams.delete(`${API_PATH}/official_feedback/${projectId}`, projectId);
}
