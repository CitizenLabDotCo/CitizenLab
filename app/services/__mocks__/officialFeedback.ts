import { IOfficialFeedback } from 'services/officialFeedback';

export const mockOfficialFeedback = {
  data: [
    {
      type: 'official_feedbacks',
      id: 'feedbackId1',
      attributes: {
        author_multiloc: {
          en: 'Testing Official Department',
        },
        body_multiloc: {
          en: 'Update: Do Not Keep Calm, please panic now. End of transmission.',
        },
        created_at: 'Some time ago',
        updated_at: 'Some shorter time ago'
      }
    },
    {
      type: 'official_feedbacks',
      id: 'feedbackId2',
      attributes: {
        author_multiloc: {
          en: 'Testing Official Department',
        },
        body_multiloc: {
          en: 'This is an important official communication from the testing department. Keep calm and read on.',
        },
        created_at: 'Some longer time ago',
        updated_at: 'Some longer shorter time ago'
      }
    }
  ]
} as IOfficialFeedback;
