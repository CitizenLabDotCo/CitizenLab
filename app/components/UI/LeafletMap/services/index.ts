import {
  setup,
  init,
  addTileLayer,
  addOnClickHandler,
  changeView,
  addOnMoveHandler,
  addOnZoomHandler,
  refitBounds,
} from './setup';

import { addLayersControl, removeControl } from './controls';
import { addLayers, removeLayers, removeLayer } from './layers';
import { addMarkersToMap, addClusterGroup } from './markers';

export default {
  setup,
  init,
  addTileLayer,
  addOnClickHandler,
  changeView,
  addOnMoveHandler,
  addOnZoomHandler,
  refitBounds,
  addLayersControl,
  removeControl,
  addLayers,
  removeLayers,
  removeLayer,
  addMarkersToMap,
  addClusterGroup,
};
