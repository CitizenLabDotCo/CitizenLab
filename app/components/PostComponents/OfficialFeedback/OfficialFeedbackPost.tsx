import React, { PureComponent } from 'react';
import { adopt } from 'react-adopt';
import styled from 'styled-components';
import { isNilOrError } from 'utils/helperUtils';

// typings
import { Locale, Multiloc } from 'typings';

// components
import OfficialFeedbackEdit from './Form/OfficialFeedbackEdit';
import MoreActionsMenu, { IAction } from 'components/UI/MoreActionsMenu';
import T from 'components/T';
import QuillEditedContent from 'components/UI/QuillEditedContent';

// styles
import { colors, fontSizes } from 'utils/styleUtils';
import { lighten } from 'polished';

// i18n
import messages from './messages';
import { FormattedMessage, injectIntl } from 'utils/cl-intl';
import { FormattedDate, InjectedIntlProps } from 'react-intl';
import { getLocalized } from 'utils/i18n';

// services
import { IOfficialFeedbackData, deleteOfficialFeedbackFromIdea, deleteOfficialFeedbackFromInitiative } from 'services/officialFeedback';

// resources
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';
import GetTenantLocales, { GetTenantLocalesChildProps } from 'resources/GetTenantLocales';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  border-radius: ${(props: any) => props.theme.borderRadius};
  color: ${colors.text};
  font-size: ${fontSizes.base}px;
  padding: 30px;
  padding-top: 35px;
  margin-bottom: 10px;

  &.last {
    margin-bottom: 0;
  }
`;

const PostContainer = styled(Container)`
  white-space: pre-line;
  background: ${lighten(0.545, colors.clRedError)};
  background: rgba(236, 90, 36, 0.06);
  position: relative;
`;

const EditFormContainer = styled(Container)`
  background: ${colors.adminBackground};
`;

const Body = styled.div`
  margin-bottom: 30px;
`;

const Footer = styled.div`
  display: flex;
  flex-direction: column;
`;

const Author = styled.span`
  color: ${colors.label};
  font-size: ${fontSizes.base}px;
  font-weight: 500;
`;

const DatePosted = styled.span`
  color: ${colors.label};
  font-size: ${fontSizes.small}px;
  font-weight: 300;
`;

const DateEdited = styled.span`
  color: ${colors.label};
  font-size: ${fontSizes.small}px;
  font-weight: 300;
  font-style: italic;
`;

const StyledMoreActionsMenu = styled(MoreActionsMenu)`
  position: absolute;
  top: 10px;
  right: 10px;
`;

interface DataProps {
  locale: GetLocaleChildProps;
  tenantLocales: GetTenantLocalesChildProps;
}

interface InputProps {
  editingAllowed: boolean | null;
  officialFeedbackPost: IOfficialFeedbackData;
  last: boolean;
  postType: 'idea' | 'initiative';
}

interface Props extends InputProps, DataProps {}

interface State {
  showEditForm: boolean;
}

export class OfficialFeedbackPost extends PureComponent<Props & InjectedIntlProps, State> {
  constructor(props: Props) {
    super(props as any);
    this.state = {
      showEditForm: false
    };
  }

  showEditForm = () => {
    this.setState({ showEditForm: true });
  }

  closeEditForm = () => {
    this.setState({ showEditForm: false });
  }

  deletePost = (postId: string) => () => {
    const { postType } = this.props;

    if (window.confirm(this.props.intl.formatMessage(messages.deletionConfirmation))) {
      switch (postType) {
        case 'idea':
          deleteOfficialFeedbackFromIdea(postId);
        case 'initiative':
          deleteOfficialFeedbackFromInitiative(postId);
        }
    }
  }

  getActions = (postId: string) => [
    {
      label: <FormattedMessage {...messages.editOfficialFeedbackPost} />,
      handler: this.showEditForm,
      name: 'edit'
    },
    {
      label: <FormattedMessage {...messages.deleteOfficialFeedbackPost} />,
      handler: this.deletePost(postId),
      name: 'delete'
    }
  ] as IAction[]

  getPostBodyText = (postBodyMultiloc: Multiloc, locale: Locale, tenantLocales: Locale[]) => {
    const postBodyText = getLocalized(postBodyMultiloc, locale, tenantLocales);
    const processedPostBodyText = postBodyText.replace(
      /<span\sclass="cl-mention-user"[\S\s]*?data-user-id="([\S\s]*?)"[\S\s]*?data-user-slug="([\S\s]*?)"[\S\s]*?>([\S\s]*?)<\/span>/gi,
      '<a class="mention" data-link="/profile/$2" href="/profile/$2">$3</a>'
    );
    return processedPostBodyText;
  }

  render() {
    const { editingAllowed, officialFeedbackPost, locale, tenantLocales, last } = this.props;
    const { showEditForm } = this.state;
    const { body_multiloc, author_multiloc, created_at, updated_at } = officialFeedbackPost.attributes;

    if (showEditForm) {
      return (
        <EditFormContainer key={officialFeedbackPost.id}>
          <OfficialFeedbackEdit
            feedback={officialFeedbackPost}
            closeForm={this.closeEditForm}
          />
        </EditFormContainer>
      );
    }

    if (!isNilOrError(locale) && !isNilOrError(tenantLocales)) {
      const formattedDate = (
        <FormattedDate
          value={created_at}
          year="numeric"
          month="long"
          day="numeric"
        />
      );

      return (
        <PostContainer key={officialFeedbackPost.id} className={`e2e-official-feedback-post ${last ? 'last' : ''}`}>
          {editingAllowed &&
            <StyledMoreActionsMenu ariaLabel={this.props.intl.formatMessage(messages.showMoreActions)} actions={this.getActions(officialFeedbackPost.id)} />
          }

          <QuillEditedContent>
            <Body>
              <div dangerouslySetInnerHTML={{ __html: this.getPostBodyText(body_multiloc, locale, tenantLocales) }} />
            </Body>
            <Footer>
              <Author>
                <T value={author_multiloc} />
              </Author>

              <DatePosted>
                {formattedDate}
              </DatePosted>

              {updated_at && updated_at !== created_at && (
                <DateEdited>
                  <FormattedMessage
                    {...messages.lastEdition}
                    values={{ date: formattedDate }}
                  />
                </DateEdited>
              )
            }
            </Footer>
          </QuillEditedContent>
        </PostContainer>
      );
    }

    return null;
  }
}

const Data = adopt<DataProps, {}>({
  locale: <GetLocale />,
  tenantLocales: <GetTenantLocales />
});

const OfficialFeedbackPostWithIntl = injectIntl<Props>(OfficialFeedbackPost);

export default (inputProps: InputProps) => (
  <Data>
    {dataProps => <OfficialFeedbackPostWithIntl {...inputProps} {...dataProps} />}
  </Data>
);
