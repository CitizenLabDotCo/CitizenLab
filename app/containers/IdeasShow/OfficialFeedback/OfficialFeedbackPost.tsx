import React, { PureComponent } from 'react';
import { adopt } from 'react-adopt';
import styled from 'styled-components';
import linkifyHtml from 'linkifyjs/html';
import { isNilOrError } from 'utils/helperUtils';

// typings
import { Locale, Multiloc } from 'typings';

// components
import OfficialFeedbackEdit from './Form/OfficialFeedbackEdit';
import MoreActionsMenu, { IAction } from 'components/UI/MoreActionsMenu';
import T from 'components/T';

// styles
import { colors, fontSizes } from 'utils/styleUtils';
import { transparentize, darken } from 'polished';

// i18n
import messages from './messages';
import { FormattedMessage, injectIntl } from 'utils/cl-intl';
import { FormattedDate, InjectedIntlProps } from 'react-intl';
import { getLocalized } from 'utils/i18n';

// services
import { IOfficialFeedbackData, deleteOfficialfeedback } from 'services/officialFeedback';

// resources
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';
import GetTenantLocales, { GetTenantLocalesChildProps } from 'resources/GetTenantLocales';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  border-radius: 3px;
  color: ${colors.text};
  font-size: ${fontSizes.base}px;
  padding: 17px 34px 27px 34px;
  margin-bottom: 10px;
`;

const PostContainer = Container.extend`
  background-color: rgba(236, 90, 36, 0.06);

  a {
    color: ${colors.clBlueDark};

    &.mention {
      background: ${transparentize(0.9, colors.clBlueDark)};
      padding-left: 4px;
      padding-right: 4px;

      &:hover {
        background: ${transparentize(0.8, colors.clBlueDark)};
      }
    }

    &:not(.mention){
      text-decoration: underline;
      overflow-wrap: break-word;
      word-wrap: break-word;
      word-break: break-all;
      word-break: break-word;
      hyphens: auto;

      &:hover {
        text-decoration: underline;
        color: ${darken(0.15, colors.clBlueDark)};
      }
    }

    &:hover {
      color: ${darken(0.15, colors.clBlueDark)};
    }
  }
`;

const EditFormContainer = Container.extend`
  background-color: ${colors.adminBackground};
`;

const Body = styled.div`
  line-height: 23px;
  margin-bottom: 16px;
  white-space: pre-line;
`;

const Footer = styled.div`
  display: flex;
  flex-direction: column;
`;

const Author = styled.span`
  font-weight: 600;
`;

const DatePosted = styled.span`
  color: ${colors.label};
`;

const DateEdited = styled.span`
  color: ${colors.label};
  font-size: ${fontSizes.small}px;
  font-style: italic;
  margin-top: 10px;
`;

const StyledMoreActionsMenu = styled(MoreActionsMenu)`
  align-self: flex-end;
  margin-bottom: 10px;
`;

interface DataProps {
  locale: GetLocaleChildProps;
  tenantLocales: GetTenantLocalesChildProps;
}

interface InputProps {
  editingAllowed: boolean | null;
  officialFeedbackPost: IOfficialFeedbackData;
}

interface Props extends InputProps, DataProps {}

interface State {
  showEditForm: boolean;
}

export class OfficialFeedbackPost extends PureComponent<Props & InjectedIntlProps, State> {
  constructor(props: Props) {
    super(props as any);
    this.state = {
      showEditForm: false
    };
  }

  showEditForm = () => {
    this.setState({ showEditForm: true });
  }

  closeEditForm = () => {
    this.setState({ showEditForm: false });
  }

  deletePost = (postId: string) => () => {
    if (window.confirm(this.props.intl.formatMessage(messages.deletionConfirmation))) {
      deleteOfficialfeedback(postId);
    }
  }

  getActions = (postId: string) => [
    {
      label: <FormattedMessage {...messages.editOfficialFeedbackPost} />,
      handler: this.showEditForm,
      name: 'edit'
    },
    {
      label: <FormattedMessage {...messages.deleteOfficialFeedbackPost} />,
      handler: this.deletePost(postId),
      name: 'delete'
    }] as IAction[]

  getPostBodyText = (postBodyMultiloc: Multiloc, locale: Locale, tenantLocales: Locale[]) => {
    const postBodyText = getLocalized(postBodyMultiloc, locale, tenantLocales);
    const processedPostBodyText = linkifyHtml(postBodyText.replace(
      /<span\sclass="cl-mention-user"[\S\s]*?data-user-id="([\S\s]*?)"[\S\s]*?data-user-slug="([\S\s]*?)"[\S\s]*?>([\S\s]*?)<\/span>/gi,
      '<a class="mention" data-link="/profile/$2" href="/profile/$2">$3</a>'
    ));
    return processedPostBodyText;
  }

  render() {
    const { editingAllowed, officialFeedbackPost, locale, tenantLocales } = this.props;
    const { showEditForm } = this.state;
    const { body_multiloc, author_multiloc, created_at, updated_at } = officialFeedbackPost.attributes;

    if (showEditForm) {
      return (
        <EditFormContainer key={officialFeedbackPost.id}>
          <OfficialFeedbackEdit
            feedback={officialFeedbackPost}
            closeForm={this.closeEditForm}
          />
        </EditFormContainer>
      );
    }

    if (!isNilOrError(locale) && !isNilOrError(tenantLocales)) {
      return (
        <PostContainer key={officialFeedbackPost.id} className="e2e-official-feedback-post">
          {editingAllowed &&
            <StyledMoreActionsMenu ariaLabel={this.props.intl.formatMessage(messages.showMoreActions)} actions={this.getActions(officialFeedbackPost.id)} />
          }

          <>
            <Body>
              <div dangerouslySetInnerHTML={{ __html: this.getPostBodyText(body_multiloc, locale, tenantLocales) }} />
            </Body>
            <Footer>
              <Author>
                <T value={author_multiloc} />
              </Author>
              <DatePosted><FormattedDate value={created_at} /></DatePosted>
              {updated_at && updated_at !== created_at && (
                <DateEdited>
                  <FormattedMessage
                    {...messages.lastEdition}
                    values={{ date: <FormattedDate value={updated_at} /> }}
                  />
                </DateEdited>
              )
            }
            </Footer>
          </>
        </PostContainer>
      );
    }

    return null;
  }
}

const OfficialFeedbackPostWithIntl = injectIntl<Props>(OfficialFeedbackPost);

const Data = adopt<DataProps, {}>({
  locale: <GetLocale />,
  tenantLocales: <GetTenantLocales />
});

export default (inputProps: InputProps) => <Data>{dataProps => <OfficialFeedbackPostWithIntl {...inputProps} {...dataProps} />}</Data>;
