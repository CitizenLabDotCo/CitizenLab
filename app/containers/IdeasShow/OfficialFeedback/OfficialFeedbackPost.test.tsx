import React from 'react';

import { shallow, mount } from 'enzyme';

import { mockOfficialFeedback } from 'services/officialFeedback';

jest.mock('services/officialFeedback');
jest.mock('./Form/OfficialFeedbackEdit');
jest.mock('utils/cl-intl');
const Intl = require('utils/cl-intl/__mocks__/');
const { intl } = Intl;

const mockOfficialFeedbackPost = mockOfficialFeedback.data[0];

import { OfficialFeedbackPost } from './OfficialFeedbackPost';

describe('<OfficialFeedbackPost />', () => {
  it('renders correctly for non-admin', () => {
    const wrapper = shallow(
      <OfficialFeedbackPost
        editingAllowed={false}
        officialFeedbackPost={mockOfficialFeedbackPost}
        intl={intl}
      />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly for admin', () => {
    const wrapper = shallow(
      <OfficialFeedbackPost
        editingAllowed={true}
        officialFeedbackPost={mockOfficialFeedbackPost}
        intl={intl}
      />);
    expect(wrapper).toMatchSnapshot();
  });
  it('when admin clicks edit it reacts in an adequate manner', () => {
    const wrapper = mount(
      <OfficialFeedbackPost
        editingAllowed={true}
        officialFeedbackPost={mockOfficialFeedbackPost}
        intl={intl}
      />);
    wrapper.find('MoreActionsMenu').find('button').simulate('click');
    wrapper.find('.e2e-action-edit').find('button').simulate('click');
    expect(wrapper.find('OfficialFeedbackEdit').length).toEqual(1);
  });
});
