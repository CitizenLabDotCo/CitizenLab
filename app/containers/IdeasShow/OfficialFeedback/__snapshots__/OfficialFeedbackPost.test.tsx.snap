// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OfficialFeedbackPost /> renders correctly for admin 1`] = `
<OfficialFeedbackPost__Container
  key="feedbackId1"
>
  <OfficialFeedbackPost__StyledMoreActionsMenu
    actions={
      Array [
        Object {
          "handler": [Function],
          "label": <FormattedMessage
            defaultMessage="Edit"
            id="app.components.IdeasShow.OfficialFeedback.editOfficialFeedbackPost"
          />,
          "name": "edit",
        },
        Object {
          "handler": [Function],
          "label": <FormattedMessage
            defaultMessage="Delete"
            id="app.components.IdeasShow.OfficialFeedback.deleteOfficialFeedbackPost"
          />,
          "name": "delete",
        },
      ]
    }
    hiddenLabel="Show more actions"
  />
  <OfficialFeedbackPost__Body>
    <T
      supportHtml={true}
      value={
        Object {
          "en": "Update: Do Not Keep Calm, please panic now. End of transmission.",
        }
      }
    />
  </OfficialFeedbackPost__Body>
  <OfficialFeedbackPost__Footer>
    <OfficialFeedbackPost__Author>
      <T
        value={
          Object {
            "en": "Testing Official Department",
          }
        }
      />
    </OfficialFeedbackPost__Author>
    <OfficialFeedbackPost__DatePosted>
      <FormattedDate
        value="Some time ago"
      />
    </OfficialFeedbackPost__DatePosted>
  </OfficialFeedbackPost__Footer>
</OfficialFeedbackPost__Container>
`;

exports[`<OfficialFeedbackPost /> renders correctly for non-admin 1`] = `
<OfficialFeedbackPost__Container
  key="feedbackId1"
>
  <OfficialFeedbackPost__Body>
    <T
      supportHtml={true}
      value={
        Object {
          "en": "Update: Do Not Keep Calm, please panic now. End of transmission.",
        }
      }
    />
  </OfficialFeedbackPost__Body>
  <OfficialFeedbackPost__Footer>
    <OfficialFeedbackPost__Author>
      <T
        value={
          Object {
            "en": "Testing Official Department",
          }
        }
      />
    </OfficialFeedbackPost__Author>
    <OfficialFeedbackPost__DatePosted>
      <FormattedDate
        value="Some time ago"
      />
    </OfficialFeedbackPost__DatePosted>
  </OfficialFeedbackPost__Footer>
</OfficialFeedbackPost__Container>
`;
